In [1]:
import numpy as np
from scipy.stats import norm
from statsmodels.stats.weightstats import ztest

$$  Z = \frac{X - \mu}{\sigma} \Rightarrow X = Z \cdot \sigma + \mu  $$

In [33]:
np.random.seed(0)
N = 100
mu = 0.2
sigma = 1
x = np.random.randn(N)*sigma + mu

In [34]:
# two-sided test
# https://www.statsmodels.org/stable/generated/statsmodels.stats.weightstats.ztest.html
print(f"z-statistic: {ztest(x)[0]:.8}\np-value:     {ztest(x)[1]:.6}")

z-statistic: 2.5648404
p-value:     0.0103223


In [35]:
# two-sided test (a pie)
mu_hat = x.mean()
sigma_hat = x.std(ddof=1)
z = mu_hat / (sigma_hat / np.sqrt(N)) # our mu0 = 0,  no pinta en la fórmula por eso
p_right = 1 - norm.cdf(np.abs(z)) # alt: norm.sf
p_left  = norm.cdf(-np.abs(z))
p = p_right + p_left
z, p

(2.564840415351368, 0.010322326848815901)

In [36]:
# one-sided test
ztest(x, alternative='larger')

(2.5648404153513686, 0.00516116342440792)

Note that z is exactly the same and the p-value is half of the two-sided p-value

In [37]:
# two-sided test (a pie)
mu_hat = x.mean()
sigma_hat = x.std(ddof=1)
z = mu_hat / (sigma_hat / np.sqrt(N)) # our mu0 = 0
p = 1 - norm.cdf(z) 
z,p

(2.564840415351368, 0.005161163424407977)

In [40]:
# null under a different reference value
mu0 = 0.2
ztest(x, value = mu0)

(0.5904283402851699, 0.5549035151647227)

Se observa un p-value muy grande (0.55), lo cual hace sentido, ya que $$Ho: \mu1=\mu2$$ es cierta... 
We fail to reject the null hipotesis

In [41]:
# Ahora a pata...
mu_hat = x.mean()
sigma_hat = x.std(ddof=1)
z = (mu_hat - mu0) / (sigma_hat / np.sqrt(N)) # mu0 <> 0
p_right = 1 - norm.cdf(np.abs(z))
p_left = norm.cdf(-np.abs(z))
p = p_right + p_left
z, p

(0.5904283402851698, 0.5549035151647228)

In [57]:
# two-sample test
N0 = 100
mu0 = 0.2
sigma0 = 1
x0 = np.random.randn(N0)*sigma0 + mu0

N1 = 100
mu1 = 0.5
sigma1 = 1
x1 = np.random.randn(N1)*sigma1 + mu1

In [58]:
ztest(x0,x1)

(-1.179278372692488, 0.23828734723243983)

In [62]:
# two-sided  a pata
mu_hat0 = x0.mean()
mu_hat1 = x1.mean()
dmu_hat = mu_hat1 - mu_hat0
s2_hat0 = x0.var(ddof=1)
s2_hat1 = x1.var(ddof=1)
s_hat = np.sqrt(s2_hat0/N0 + s2_hat1/N1)
z = dmu_hat / s_hat # reference value is 0
p_right = 1 - norm.cdf(np.abs(z))
p_left = norm.cdf(-np.abs(z))
p = p_right + p_left
z, p

(1.179278372692488, 0.23828734723243983)

The statistic may be different by a sign, this happens becoause the test may exchange x0 and x1.

In [63]:
# show that we will reject the null hypotesis when the 
# null hypothesis is true (false alarm) 5% of the time

num_tests= 10000
results = np.zeros(num_tests)
for i in range(num_tests):
    x1 = np.random.rand(100)
    x2 = np.random.rand(100)
    z, p = ztest(x1, x2)
    results[i] = (p < 0.05)
print(results.mean())

0.0522


Esto significa que las medias sí son iguales el 95% de las veces, o mejor dicho, el 5.22% de las veces, no logramos rechazar la hipótesis de que las medias son iguales